### PR TITLE
Core / ZeroCopy: Added Doxygen API documentation and 2 new c++ samples demonstrating usage of low level zero copy API

### PR DESCRIPTION
**Pull request by rex-schilasky**
**Pull request type**

Please check the type of change your PR introduces:
- [X] Feature
- [X] Documentation content changes

**What is the current behavior?**
We missed samples to demonstrate low level zero copy API and the CPayloadWriter class was not documented (Doxygen).

Issue Number: #1145

**What is the new behavior?**
Two new binary send / rec zero copy samples added, CPayloadWriter class documented.

**Does this introduce a breaking change?**

- [ ] Yes
- [X] No

**Cherry-Pick to**
- 5.12

### CODE REVIEW

**Comment by FlorianReimold** [comment]
If I see correctly, this PR basically contains 2 changes:

1. It adds the samples
2. More importantly: It adds doxygen documentation to the payload writer (public API)

The latter change requires this commit to be cherry-picked to 5.12.

@rex-schilasky can you confirm this?

**Comment by rex-schilasky** [comment]
> If I see correctly, this PR basically contains 2 changes:
> 
> 1. It adds the samples
> 2. More importantly: It adds doxygen documentation to the payload writer (public API)
> 
> The latter change requires this commit to be cherry-picked to 5.12.
> 
> @rex-schilasky can you confirm this?

This is all correct. The doxygen documented `CPayloadWriter` class and both samples demonstrating the usage of this class should go to 5.12.

**Review by FlorianReimold — CHANGES_REQUESTED** [review]

File: ecal/core/include/ecal/ecal_payload_writer.h [hunk 1]
```diff
@@ -29,56 +29,140 @@
...
+     * @return The size of the memory that needs to be copied.
+     */
     size_t GetSize() override { return m_size; };
-  
+
   private:
-    const void* m_buffer = nullptr;
-    size_t      m_size   = 0;
+    const void* m_buffer = nullptr;  ///< Pointer to the buffer containing the data to be written.
+    size_t      m_size = 0;          ///< Size of the data to be written.
```
Comment (FlorianReimold) [inline ecal/core/include/ecal/ecal_payload_writer.h:165]: Is this really the "data to be written"? If so, why isn't this just a temporary locally scoped variable? Or is this capacity of the memory file, as the `GetSize()` function says?

Either way, I guess either the documentation of the variable or the `GetSize()` function is wrong

**Review by FlorianReimold — CHANGES_REQUESTED** [review]
After reviewing the Doxygen I don't really get what "data to be written" means here. I assume that this term refers to the size of the underlying memory file?

File: ecal/core/include/ecal/ecal_payload_writer.h (hunk 1 above)
Comment (rex-schilasky) [reply ecal/core/include/ecal/ecal_payload_writer.h:165]: The `CBufferPayloadWriter` is a kind of primitive wrapper of the classic `CBufferPayloadWriter(const void* const buffer_, size_t size_)` Send API. Basically it stores the arguments `buffer_` and `size_` into the internals of `CBufferPayloadWriter` and does the memcpy action in the `Write` call.
`GetSize` reports the needed size correctly, Write is just doing the copy.

**Review by FlorianReimold — APPROVED** [review]

**Pull request merged**